### DISCUSSION TITLE
0086 XLS-86d: Firewall

### DISCUSSION THREAD

**Discussion by xrpl365** [XRPLF/XRPL-Standards#255] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by tequdev** [comment]
Very interesting proposal!

1. Could the sfBackupAccount be the default WhiteLited account and the account set in the FirewallWhilelistSet be considered a second Backup account? (so this spec allows a minimum of 1 maximum of 2 Backup accounts?)

2. This only allows Native Amount, why not support IOU Amount?

3. It may need to be checked that the sfPublicKey in FirewallSet is not a RegularKey for the Account. And possibility, it is also necessary to fix not to set with SetRegularKey to the account associated with the sfPublicKey in FirewallWhitelist. (Also, this could be the case with SignerListSet as well...)

4. Path Payment can be an attack vector even if a Firewall is set up at the time of the private key compromise. An attacker can set arbitrary malicious sfPaths in Path Payment to a WhiteListed account.

**Reply by dangell7** [reply]
1. Not exactly sure what you're getting at. The backup is really the default whitelisted. At first, I just used the sfBackup to add that to the whitelist. I might revert to that as far as implementation purposes but the point is. You cannot add a firewall without a default whitelist. The sfBackup just makes it SUPER clear which account that is.

2. We want to support IOU amounts but the trick is that with many (1000's) IOUs it could become cumbersome. So putting it on the Trustline would be my initial thought. MPTs now add an additional complexity. 

3. Excellent Point!

4. Good Point! Open to suggestions. Disabling the ability to use pathing when firewall is enabled would be my first instinct. Can you make any recomendations?

**Reply by tequdev** [reply]
I guess it depends on how you limit it in OfferCreate. If you want to specify the amount of XRP that can be sold, I think you can do the same with Swap (Self Payment).

However, restrictions on transaction types which do not have Destination may not be suitable for daily use of the account.

**Comment by mvadari** [comment]
1. What's the objective in having a `BackupAccount` _and_ a firewall whitelist?
2. IMO this should support multisig for `sfPublicKey`.
3. Why can't an account be deleted with a `Firewall` enabled if the `Destination` is on the whitelist?
4. IMO `PublicKey` and `Signature` are a bit too general of field names for this, and could be confused with `SigningPubKey` and `TxnSignature`.
5. Ditto with `Amount` - before I fully read the spec, I thought that the way this would be implemented is that every transaction over the `Amount` value would be "held" for the `TimePeriod`. Perhaps `TotalAmount` would be better?

**Reply by dangell7** [reply]
1. I'm removing that and just doing the "whitelist". 
2. If we did that the entire account would need multisig and that is what we are trying to avoid. 
3. It can be (Will Update)
4. Noted will add specific fields for these
5. Noted will update. This is more like a limit you place on your bank account. Cannot send more than $500 in 24 hours.

**Reply by xrpl365** [reply]
> 1. What's the objective in having a `BackupAccount` _and_ a firewall whitelist?
Essentially they are one of the same, you could think of the backupAccount as being slot 0.
Denis and I did discuss just implementing it as such, however, we MUST ensure that slot 0 is set as a required field.
This decision I leave to Denis and the reviewers to go for the most efficient solution.

> 2. IMO this should support multisig for `sfPublicKey`.
Keen to hear more from you on this.

> 3. Why can't an account be deleted with a `Firewall` enabled if the `Destination` is on the whitelist?
Well, it could be, we disabled delete to stop the exploit of using destination, so again my view is that we would be flexible and go with the most efficient solution.

> 4. IMO `PublicKey` and `Signature` are a bit too general of field names for this, and could be confused with `SigningPubKey` and `TxnSignature`.
Totally happy to take advice on naming conventions.

> 5. Ditto with `Amount` - before I fully read the spec, I thought that the way this would be implemented is that every transaction over the `Amount` value would be "held" for the `TimePeriod`. Perhaps `TotalAmount` would be better?
This one I think we should discuss offline.

**Reply by mvadari** [reply]
> If we did that the entire account would need multisig and that is what we are trying to avoid.

Why? Only exceeding the firewall amount should require that multisig in that case.

> > IMO this should support multisig for sfPublicKey.

> Keen to hear more from you on this.

A corporate account may want additional protections on their account, to ensure that it can't be hacked. For example, a bank may want to have additional approvals in order to send a larger amount of money somewhere. This principle could apply to issuers or other projects too. 


6. Why use a public key instead of an account, like with `RegularKey` and `Signers`?

**Reply by xrpl365** [reply]
> 6\. Why use a public key instead of an account, like with `RegularKey` and `Signers`?

Acknowledged and addressed in next version

**Comment by xrpl365** [comment]
I was thinking we should also add protection against malicious actors.
If they can't get your XRP, maybe they burn it, so you can't have it either.
Need protection against setting a deliberate fee to drain your xrp.
@dangell7 can we add in fee protection ?

**Reply by xrpl365** [reply]
This will be addressed in next version of the spec.

**Comment by xrpl365** [comment]
This discussion has now been closed and transitioned to: https://github.com/XRPLF/XRPL-Standards/pull/363
